core/src/consensus/tendermint/vote_regression_checker.rs

```diff
@@ -0,0 +1,189 @@
+    #[test]
+    fn test_disallow_height_decrease() {
+        let mut checker = VoteRegressionChecker::new();
+
+        checker.check(&VoteOn {
+            step: VoteStep::new(100, 10, Step::Prevote),
+            block_hash: Some(H256::random()),
+        });
+
+        assert!(!checker.check(&VoteOn {
+            step: VoteStep::new(99, 10, Step::Prevote),
+            block_hash: Some(H256::random())
+        }))
+    }
+
+    #[test]
+    fn test_allow_view_increase() {
+        let mut checker = VoteRegressionChecker::new();
+
+        checker.check(&VoteOn {
+            step: VoteStep::new(100, 10, Step::Prevote),
+            block_hash: Some(H256::random()),
+        });
+
+        assert!(checker.check(&VoteOn {
+            step: VoteStep::new(100, 11, Step::Prevote),
+            block_hash: Some(H256::random())
+        }))
+    }
+
+    #[test]
+    fn test_disallow_view_decrease() {
+        let mut checker = VoteRegressionChecker::new();
+
+        checker.check(&VoteOn {
+            step: VoteStep::new(100, 10, Step::Prevote),
+            block_hash: Some(H256::random()),
+        });
+
+        assert!(!checker.check(&VoteOn {
+            step: VoteStep::new(100, 9, Step::Prevote),
+            block_hash: Some(H256::random())
+        }))
+    }
+
+    #[test]
+    fn test_allow_step_increased() {
+        let mut checker = VoteRegressionChecker::new();
+
+        checker.check(&VoteOn {
+            step: VoteStep::new(100, 10, Step::Prevote),
+            block_hash: Some(H256::random()),
+        });
+
+        assert!(checker.check(&VoteOn {
+            step: VoteStep::new(100, 10, Step::Precommit),
+            block_hash: Some(H256::random())
+        }))
+    }
+
+    #[test]
+    fn test_disallow_step_decreased() {
+        let mut checker = VoteRegressionChecker::new();
+
+        checker.check(&VoteOn {
+            step: VoteStep::new(100, 10, Step::Prevote),
+            block_hash: Some(H256::random()),
+        });
+
+        assert!(!checker.check(&VoteOn {
+            step: VoteStep::new(100, 10, Step::Propose),
+            block_hash: Some(H256::random())
+        }))
+    }
+
+    #[test]
+    fn test_allow_same_hash() {
+        let mut checker = VoteRegressionChecker::new();
+
+        let block_hash = Some(H256::random());
+        checker.check(&VoteOn {
+            step: VoteStep::new(100, 10, Step::Prevote),
+            block_hash,
+        });
+
+        assert!(checker.check(&VoteOn {
+            step: VoteStep::new(100, 10, Step::Prevote),
+            block_hash,
+        }))
+    }
+
+    #[test]
+    fn test_disallow_hash_change() {
+        let mut checker = VoteRegressionChecker::new();
+
+        checker.check(&VoteOn {
+            step: VoteStep::new(100, 10, Step::Prevote),
+            block_hash: Some(H256::random()),
+        });
+
+        assert!(!checker.check(&VoteOn {
+            step: VoteStep::new(100, 10, Step::Prevote),
+            block_hash: Some(H256::random()),
```

Comment:
We need to make sure that this hash is a different value from the previous one.

Reply:
I've hard-coded values